src/porepy/examples/geothermal_reservoir.py

```diff
@@ -0,0 +1,493 @@
+        add_nonlinear_diffusive_flux_discretization: Callable[
+            [pp.ad.MergedOperator], None
+        ]
+        remove_nonlinear_diffusive_flux_discretization: Callable[
+            [pp.ad.MergedOperator], bool
+        ]
+
+    def bc_type_darcy_flux(self, sd: pp.Grid) -> pp.BoundaryCondition:
+        """Return Neumann BC for Darcy flux on well grids during first time interval.
+
+        Parameters:
+            sd: The subdomain for which to return the BC type.
+
+        Returns:
+            The boundary condition type for Darcy flux on the given subdomain.
+        """
+        if (
+            self.is_well_grid(sd)
+            and self.time_manager.time <= self.time_manager.schedule[1]
+        ):
+            # Before start of injection, impose Neumann BCs on well grids. A zero-flux
+            # condition is imposed by default when no BC values are specified. The <=
+            # comparison ensures that the BCs are kept as Neumann as long as the time
+            # step is within the first time interval [0, t1], where t1 is the start time
+            # of injection, consistent with the implicit time stepping employed in
+            # PorePy.
+            domain_sides = self.domain_boundary_sides(sd)
+            inds = domain_sides.top if self.nd == 3 else domain_sides.north
+            return pp.BoundaryCondition(sd, inds, "neu")
+        else:
+            return super().bc_type_darcy_flux(sd)  # type: ignore[misc]
+
+    def bc_type_fourier_flux(self, sd: pp.Grid) -> pp.BoundaryCondition:
+        """Return Neumann BC for Fourier flux on well grids during first time interval.
+
+        Parameters:
+            sd: The subdomain for which to return the BC type.
+        Returns:
+            The boundary condition type for Fourier flux on the given subdomain.
+        """
+        if (
+            self.is_well_grid(sd)
+            and self.time_manager.time <= self.time_manager.schedule[1]
+        ):
+            # Before start of injection, impose Neumann BCs on well grids. A zero-flux
+            # condition is imposed by default when no BC values are specified. The <=
+            # comparison ensures that the BCs are kept as Neumann as long as the time
+            # step is within the first time interval [0, t1], where t1 is the start time
+            # of injection, consistent with the implicit time stepping employed in
+            # PorePy.
+            domain_sides = self.domain_boundary_sides(sd)
+            inds = domain_sides.top if self.nd == 3 else domain_sides.north
+            return pp.BoundaryCondition(sd, inds, "neu")
+        else:
+            return super().bc_type_fourier_flux(sd)  # type: ignore[misc]
+
+    def before_nonlinear_loop(self) -> None:
+        """Hook called before the nonlinear solver loop.
+
+        Rediscretize flow problem at the onset of second time interval, when injection
+        starts. The tailoring is done to recompute the Darcy flux discretization only at
```

Comment:
I don't see the motivation for this operation; the time saved will be miniscule, right?

Reply:
It feels like a more "proper" way of doing the rediscretization. Let's discuss.